source/benchmarking/odm-benchmarking.md

```diff
@@ -0,0 +1,385 @@
+
+Dataset size: For scoring purposes, the dataset size is the size of the `large_doc_nested` source file (8,000 bytes)
+times 10,000 operations, which equals 80,000,000 bytes or 80 MB.
+
+| Phase       | Description                                                                                                                                                                                                                                                  |
+| ----------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
+| Setup       | Load the LARGE_DOC_NESTED dataset into memory as an ODM-appropriate model object. Insert 10,000 instances into the database, saving the value of the `unique_id` field for the first item in each model's `embedded_str_doc_array` nested model into a list. |
+| Before task | n/a.                                                                                                                                                                                                                                                         |
+| Do task     | For each of the 10,000 `unique_id` values, perform a filter operation to search for the parent LARGE_DOC_NESTED model.                                                                                                                                       |
+| After task  | n/a.                                                                                                                                                                                                                                                         |
+| Teardown    | Drop the collection associated with the LARGE_DOC_NESTED model.                                                                                                                                                                                              |
+
+## Benchmark platform, configuration and environments
+
+### Benchmark Client
+
+The benchmarks should be run with the most recent stable version of the ODM and the newest version of the driver it
+supports.
+
+### Benchmark Server
+
+The MongoDB ODM Performance Benchmark must be run against a MongoDB replica set of size 1 running the latest stable
+database version without authentication or SSL enabled.
+
+### Benchmark placement and scheduling
+
+The MongoDB ODM Performance Benchmark should be placed within the ODM's test directory as an independent test suite. Due
+to the relatively long runtime of the benchmarks, including them as part of an automated suite that runs against every
+PR is not recommended. Instead, scheduling benchmark runs on a regular cadence is the recommended method of automating
+this suite of tests.
+
```

Comment:
Do we have a dedicated, isolated perf lab, with machines that won't get changes unless we know about it? My experience with perf testing over many years is that unless you have such a system, then the noise makes it very difficult to see when things really change. For example, OS updates, platform/language changes, virus checking, auto-updates kicking in mid run, and so on, all make the data hard to interpret.

How do you currently handle driver perf test machines? Can you point me to charts, or even raw data I guess, that should variation/noise over time? Also, how often do they run? Is there only a single change between each run so that it's feasible to trace back a perf difference to a single change, be that external or a code change?

Reply:
Drivers do have access to a special host in evergreen to run dedicated performance tasks on to ensure stability and consistency (`rhel90-dbx-perf-large`).